text/3809-derive-from.md

```diff
@@ -0,0 +1,370 @@
+
+Based on the popularity of the `derive_more` crate (discussed in [Prior art](#prior-art)), which had more than 125 million downloads when this RFC was proposed, it seems that there is a lot of appetite for extending the set of use-cases where deriving standard traits is allowed. This feature was discussed in the past [here](https://github.com/rust-lang/rfcs/issues/2026).
+
+The proposed change enables the usage of an existing feature (`#[derive]`) in more situations. It makes code easier to read by using an intuitive built-in feature instead of forcing users to write boilerplate code or use macros.
+
+As always, an alternative is to just not do this, in that case users would continue implementing `From` using boilerplate code, macros or external crates.
+
+Because the scope of the proposed change is quite minimal, it should be forward-compatible with designs that would make it work in more situations in the future (some ideas are discussed in [Future possibilities](#future-possibilities)). There is one potential (although unlikely) incompatibility discussed below.
+
+## Alternative design using tuples
+There is one possible alternative design that comes to mind which could be in theory incompatible with this proposal. We could enable `#[derive(From)]` for tuple structs with an arbitrary number of fields by generating a `From` implementation from a tuple containing the types of the struct fields:
+
+```rust
+#[derive(From)]
+struct Foo(u32, u16, bool);
+
+impl From<(u32, u16, bool)> for Foo { ... }
+```
+
+The question then becomes what would be generated under this design when the struct has exactly one field.
+
+- We could either generate `From<T> for Type`, which would be compatible with this RFC. It would also be slightly inconsistent though, as it would generate something different only for the case with a single field.
+    - This is how the `derive_more::From` macro behaves.
+- Or, we could generate `From<(T, )> for Type`, which would be consistent with the logic of generating `From<tuple>`. However, single-field tuples are not idiomatic and it would be awkward having to write e.g. `(value, ).into()` to make use of the impl. 
+
+I think that the second approach is not a good idea, and I find it unlikely that we would want to use it.
+
+## Generating `From` in the other direction
+This proposed change is useful to generate a `From` impl that turns the inner field into the wrapper struct (`impl From<Inner>` for `Newtype`). However, sometimes it is also useful to generate the other direction, i.e. turning the newtype back into the inner type. This can be implemented using `impl From<Newtype> for Innertype`.
+
+We could make `#[derive(From)]` generate both directions, but that would make it impossible to only ask for the "basic" `From` direction without some additional syntax.
+
+A better alternative might be to support generating the other direction in the future through something like `#[derive(Into)]`.
```

Comment:
It might be worth noting here that `impl From<Newtype> for Inner` and `impl Into<Inner> for Newtype` have slightly different semantics. Using `derive(Into)` to mean `impl From` could be confusing for generic types where there is a coherence issue, even if it does imply an `Into` impl.

(I've had similar issues with the derive_more version.)

Reply:
the derive-macro's name does not necessarily correspond to the impl'ed trait name anymore since #3621 (currently named `#[derive(CoercePointee)]`, which actually does `impl CoerceUnsized + DispatchFromDyn`.)

Reply:
The `#[derive(Into)]` example is kind of hand-waving, I'm not sure if it's actually a good idea. I would like to avoid doing that in this RFC though, as that sounds like a separate can of worms, I explicitly tried to keep this RFC as simple as possible.

Reply:
Honestly, for the `impl From<NewType> for Inner` case, I think I would prefer something like `impl_from!(Newtype -> Inner)` or a more generic  `impl_from!(Newtype -> Inner = |source| source.0)` or `impl_from!(Newtype -> Inner = self.0)` (restricted to using fields of NewType)

Although it is a little annoying, to have to repeat the type of Inner.

Reply:
> But my comment was more about the confusing semantics

I don't think there is any name more appropriate than `#[derive(Into)]` especially if *this RFC* is using `#[derive(From)]`. The final effect you get is still having an `impl Into<Inner> for Self` effectively, through the intermediate `impl From<Self> for Inner` + the blanket impl. 

See JelteF/derive_more#13 for a brief discussion how [derive_more](https://docs.rs/derive_more/latest/derive_more/derive.Into.html) still chooses to name it `#[derive(Into)]`.

Reply:
I'd love to have a derive in the other direction, but full support for making that future work and not part of *this* RFC.